### ISSUE TITLE
Significant slow performance when running more than 1 thread

### ISSUE THREAD

**Issue by idanmashaal** [Badgerati/Pode#1033]
Hi, @Badgerati

Apologies for duplicating this question from #1031 ,but I'm experiencing something similar and more extreme.

I have a module that I'm using inside of Pode. Normal run of this module via CLI takes ~1500ms.
Module returns string output, I wrap that with `<pre>` inside Pode and timing is the same.
If I run this inside Pode with 1 thread, I get the same time, which is around ~1600ms.
If I use Pode with 2 threads, something interesting happens:
First request will take ~1600ms
Second request will take ~9000ms
Third request will take ~1600ms
Fourth request will take ~9000ms
and so forth....
That's around x6 times slower.

If I add more than 2 threads, the first request will be fast, then all other requests will be slow until served from the first thread.
I added timing inside my external module and can verify that the slowness comes from the reason Pode is calling the module.
Is there something that can be done to optimize this, other than running multiple instances of Pode with 1 thread on a given machine with a load-balancer?

Thanks,
Idan.

**Comment by Badgerati** [comment]
Hi @idanmashaal,

I've seen similar behaviour to this before, and it's usually module loading issue - nothing do do with the module you're using, just a PowerShell quirk.

Which module is it you're using, or just a private one? Are you able to give an example script for the Pode server of where the module is imported/loaded and used.

Usually this happens when relying on PowerShell to auto-load a module without actually doing `Import-Module` - it tries to do this every time in secondary runspaces.

**Comment by idanmashaal** [comment]
Hi @Badgerati ,

I've written a script that uses the PowerShell Module [Indented.Net.Dns](https://www.powershellgallery.com/packages/Indented.Net.Dns/6.1.2). To make my life easier, I converted my script to a module (the quick and dirty way without a psd1 file) and Import my own psm1 file in Pode's server.ps1.
I also imported Indented.Net.Dns with the Global flag, that didn't help.
I tried adding the Import-Module inside my own Module (that was a script a moment ago) but that didn't do much, so opted to remove the Import-Module call from inside my script and rely on the Pode one. That didn't help either.
Inside my script, I call Indented.Net.Dns several times, and I was able to conclude that each call does a re-import. 

Any idea how I can work around this PowerShell quirk, it's going to be quite dirty to embed all my script's logic inside Pode.

Happy to create a stripped down version of my module that will only call Indented.Net.Dns several times and share it here if this can help us figure it out.

Thanks,
Idan.

**Comment by idanmashaal** [comment]
Updates @Badgerati 

Attempt 1: Added "Import-Module Indented.Net.Dns -Global -Verbose" to my own script and every-time Pode calls it, there is an import. Much slower than before. I know it's loaded every time because the Verbose script prints to the console.
Attempt 2: Instead, I copied my script inside Pode's server.ps1 as a function, then I kept the single import module with -Global and -Verbose right before inside Start-PodeServer, still same results. Module is loaded only once by looking at the Verbose of the console.

Because Attempt 2 takes away my wrapper script but still yields the same exact results, my educated guess here would be that when running in a secondary runspace, every call to Indented.Net.Dns loads it even if I'm in the same runspace and function context. This is because every call to Indented.Net.Dns takes ~1.5 to ~2.0 seconds and I compared this with a cold import without Pode (a simple script).

I would be "okay" if the import would happen once per runspace and then it would be an average of 2 seconds per execution, that would be ~2x slower. But every call takes 2 seconds, and there are more than once, that's becoming more like 10x when using any setting of threads other than 1.

So this PowerShell quirk is a real mind blower and would be happy if we can think of a way to overcome it (if at all possible).

Thanks,
Idan.

**Comment by Badgerati** [comment]
Hi @idanmashaal,

I just tried a quick test myself, and if I import the DNS module before `Start-PodeServer`, it's imported into all other runsapces, and calls to say `Get-Dns` don't try to re-import:

```powershell
# importing here helps solve the quirk, as Pode will auto-import into all runspaces
Import-Module Indented.Net.Dns

Start-PodeServer -Thread 2 -Verbose {
    Add-PodeEndpoint -Address * -Port 8090 -Protocol Http

    Add-PodeRoute -Method Get -Path '/' -ScriptBlock {
        $start = [datetime]::Now
        "Thread: $($ThreadId)" | Out-Default
        Get-Module | Out-Default
        Get-Dns | Out-Default
        "Duration: $(([datetime]::Now -$start).TotalSeconds)" | Out-Default
        '- - - - - - - - - - - - - -' | Out-Default
    }
}
```

Few questions:
* Which OS are you using?
* Which version of PowerShell and Pode?
* Are you running this on a physical machine, VM, container, or a cloud resource?
* If you are able to show a cut down version of the server/route/script that would help :)

**Comment by idanmashaal** [comment]
Hi @Badgerati ,

I'm using macOS 12.6.1 and Ubuntu Linux 22.04.
Powershell 7.3.0 and Pode 2.7.2 on both. 
I develop on my macOS and deploy on a Digital Ocean.

My Mac Mini is Late 2018 (3.2 GHz 6-Core Intel Core i7) --> No VM's on my macOS.

I experience the same behavior on my Mac Mini and on a VM.


I am doing exactly as you are doing by importing the Indented.Net.Dns before Start-PodeServer and also tried with the -Global switch.

**Comment by idanmashaal** [comment]
The slowness actually happens when invoking Get-DNS commands in any thread that is not the first one.

Example:
$domain = "github.com"
$dns_server = "8.8.8.8"
$dns_timeout_sec = 3
$mx_records = Get-Dns -Name ($domain + ".") -RecordType MX -ComputerName $dns_server -Timeout $dns_timeout_sec -ErrorAction Stop

So for example a text box that reads the domain name, and prints back the Get-DNS value, will be slow for subsequent requests if thread is >1

**Comment by idanmashaal** [comment]
See the following code example I added on top of yours, pay special attention to the "Duration_ms" property I added.
You will see the variance between runs.
Example run will be:

http://127.0.0.1:8090/?domain=github.com

```
# importing here helps solve the quirk, as Pode will auto-import into all runspaces
Import-Module -Name Pode -MaximumVersion 2.99.99 -Verbose
Import-Module Indented.Net.Dns -Verbose
Start-PodeServer -Thread 2 -Verbose {
    Add-PodeEndpoint -Address * -Port 8090 -Protocol Http

    Add-PodeRoute -Method Get -Path '/' -ScriptBlock {
        $start = [datetime]::Now
        "Thread: $($ThreadId)" | Out-Default
        Get-Module | Out-Default
        Get-Dns | Out-Default
        "Duration: $(([datetime]::Now -$start).TotalSeconds)" | Out-Default
        '- - - - - - - - - - - - - -' | Out-Default

        $result = ''
        if ( $WebEvent.Query.Count -gt 0 )
        {
            $domain = $WebEvent.Query['domain']

            if ( ($null -ne $domain) -and ($domain.Length -gt 0) )
            {
                $dns_server = "8.8.8.8"
                $dns_timeout_sec = 3
                $timer = [System.Diagnostics.Stopwatch]::StartNew()
                $mx_records = Get-Dns -Name ($domain + ".") -RecordType MX -ComputerName $dns_server -Timeout $dns_timeout_sec -ErrorAction Stop
                $timer.Stop()
                $duration_ms = $timer.Elapsed.TotalMilliseconds
                $result = $mx_records | ConvertTo-Json -Depth 10 | ConvertFrom-Json
                $result | Add-Member -MemberType NoteProperty -Name 'Duration_ms' -Value $duration_ms
                Set-PodeHeader -Name 'Server-Timing' -Value ("Get-Dns;dur="+$duration_ms)
            }
            else 
            {
                $result = @{ "Error" = "No domains provided in query string 'domains' - use comma separated values" }
            }
        }
        $result.GetType()
        Write-PodeJsonResponse -Value ($result | ConvertTo-Json -Depth 10)
   }
}
```

**Comment by idanmashaal** [comment]
Adding screenshot that shows the difference between Thread1 and Thread2, this is the same behavior between all threads that are not Thread1.

![image](https://user-images.githubusercontent.com/377019/204145029-452dd9cc-a09c-48e0-bcc0-cd16c668533d.png)

**Comment by Badgerati** [comment]
Hi @idanmashaal,

Thanks for the info!

I took a look at the DNS module itself, and can see why this is occurring now: it's using PowerShell classes. Because of the way classes were written in PowerShell, they can only ever be updated on the thread/runspace they were originally created, meaning secondary runspaces will always delegate back to the primary. You can see this below where I forced the first call on Thread 1 to skip, causing Thread 2 to become the primary and Thread 1 be slow to respond:

![image](https://user-images.githubusercontent.com/1483395/204148158-aa035c47-635e-4815-98cd-b981c8c41ac4.png)

There is a "work-around" for this ([seen here](https://badgerati.github.io/Pode/Getting-Started/KnownIssues/#powershell-classes)), but it only works if you're the one creating the classes, rather than a module, which means unfortunately there isn't a solution for this 😢 (or at least, not one I'm aware of!)

You could use Pode's Tasks to funnel requests through Thread 1, I'm not sure how this might look on your side with a bigger example, but it will cause a block on a Task between the web runspaces:

```powershell
Import-Module -Name Pode -MaximumVersion 2.99.99 -Verbose
Import-Module Indented.Net.Dns

Start-PodeServer -Thread 2 -Verbose {
    Add-PodeEndpoint -Address * -Port 8090 -Protocol Http

    Set-PodeTaskConcurrency -Maximum 1
    Add-PodeTask -Name 'GetDns' -ScriptBlock {
        param($domain, $server)
        $timeout_sec = 3
        $mx_records = Get-Dns -Name ($domain + ".") -RecordType MX -ComputerName $server -Timeout $timeout_sec -ErrorAction Stop
        return $mx_records
    }

    Add-PodeRoute -Method Get -Path '/' -ScriptBlock {
        $result = ''
        if ( $WebEvent.Query.Count -gt 0 )
        {
            $domain = $WebEvent.Query['domain']

            if ( ($null -ne $domain) -and ($domain.Length -gt 0) )
            {
                $timer = [System.Diagnostics.Stopwatch]::StartNew()

                $mx_records = Invoke-PodeTask -Name 'GetDns' -Wait -ArgumentList @{
                    domain = $domain
                    server = '8.8.8.8'
                }

                $timer.Stop()
                $duration_ms = $timer.Elapsed.TotalMilliseconds
                $result = $mx_records | ConvertTo-Json -Depth 10 | ConvertFrom-Json
                $result | Add-Member -MemberType NoteProperty -Name 'Duration_ms' -Value $duration_ms
                $result | Add-Member -MemberType NoteProperty -Name 'Thread' -Value $ThreadId
                Set-PodeHeader -Name 'Server-Timing' -Value ("Get-Dns;dur="+$duration_ms)
            }
            else 
            {
                $result = @{ "Error" = "No domains provided in query string 'domains' - use comma separated values" }
            }
        }
        $result.GetType()
        Write-PodeJsonResponse -Value ($result | ConvertTo-Json -Depth 10)
   }
}
```

When I did the above all Threads took the same time - running requests synchronously.

**Comment by idanmashaal** [comment]
Great, I'll look at the Pode-Task approach.
Do you have an estimate on what to expect when funnling these requests via Thread 1 while having more than 1 thread for Pode-Server itself ? i.e will this improve concurrency in some way?

Other way I thought was creating a PowerShell script for as a wrapper and calling it from each runspace, but this has an overhead of a new pwsh process which takes ~2 seconds per run.

Thanks,
Idan.

**Comment by idanmashaal** [comment]
Hi @Badgerati,

I did some more testing with ThreadCount=4 and TaskConcurrency=1 and throughput increased.
Then I decided to test with ThreadCount=4 and TaskConcurrency=4 and throughput increased more, at the same time there is no latency in calls to Get-DNS as before.

So this solves the issue in some magical way! (any idea why?)

Thanks,
Idan.

**Comment by Badgerati** [comment]
Hi @idanmashaal,

Ooo, this could be because the Task runspaces are short lived, unlike the Web ones! So the primary runspace for Tasks is always being reset; if you invoke the Task enough times I bet you'll see a slow delegation call on the odd occasion.

Glad it's solved your issue! 😄

**Comment by idanmashaal** [comment]
Hi @Badgerati 
Apparently they are short lived, I tested with Loader.io and was able to reach 250 clients per minute per instance. I have two DigitalOcean droplets, each have 1vCPU and 1GB of RAM behind a load-balancer. That's a total of 500 clients per minute when Threads is 4 and TaskConcurrency is 4.
When I had 4 Threads and TaskConcurrency I reached just 150 clients per minute per instance (total of 300) --> so this is a very nice boost in performance, while the average requests is less than 2 second.
This is way more than the usage of this simple service, but shows nicely the potential.

![image](https://user-images.githubusercontent.com/377019/204396215-64760bf7-3e57-498c-9d83-7fc7803b3501.png)

We can close this question now. Thanks a lot !
